Prettify lm_eval --tasks list

anthony-dipofi opened the pull request:
As requested by @haileyschoelkopf I am working on prettifying the output from `lm_eval --tasks list` and I wanted to start a draft PR to get feedback on my approach. 

Following along with the suggestions at https://github.com/EleutherAI/lm-evaluation-harness/pull/1901#discussion_r1621274760 I created additional properties `task_manager.all_groups` and `task_manager.all_subtasks` to hold just the groups and nongroup tasks respectively. I also moved the logic for adding the newlines, etc. into a new TaskManager.list_all_tasks() method that returns a string with the list of groups, subtask or both.

I also added additional arguments so that you can use `lm_eval --tasks list_groups` to get a list of the task groups or `lm_eval --tasks list_subtasks` to get a list of the subtasks, in addition to `lm_eval --tasks list` which gets both.

So the main things I wanted to get feedback on is whether or not this overall approach is acceptable. Also in terms of rendering it to a markdown table how exactly should this be approached? Would there be other columns besides the name, and if so what should they be? And also how to handle things like various column sizes, etc. if the goal is to display the markdown table in the console. And just in general if there are any other suggestions for how to improve the readability of the output beyond just having a big list.

Thanks, please let me know any suggestions or further changes that are needed!

haileyschoelkopf reviewed (commented):
This approach looks like what I had in mind!!

For making a prettier table print, how about something along these lines, very similar to `utils.make_table():`

```
|    Task      |         Config Location          |
|--------------|----------------------------------|
| tasknameA    | lm_eval/tasks/path/to/taskA.yaml |
```

(With the `Task` header reading `Groups` for a groups table)

When we run `list_all_tasks()`, I think that groups and subtasks should also still be distinguished from each other.

The reason not to print more info is because actually loading every task config / object to get info such as the number of test-set samples requires too much overhead and requires downloading every dataset in the sample-count case.

If it's easy to also statically grab the `output_type` of a task though, then a column for that too would be the most asked-for feature.

anthony-dipofi commented:
So I added in the markdown table, with the Config Location column. 

I also added an Output Type column for the tasks table. I am not sure if this is what you had in mind but I ended up getting the output type by loading in the yaml config file for each task and reading it from there, or if no output type was specified I would read it from the yaml file that was marked as `include`. I couldn't really figure out a way to get the output type information inwithout reading in those yaml files though.

haileyschoelkopf commented:
Thank you for working on this, @anthony-dipofi ! 

I added logic to now also print a table of `tag` type values, and will merge!
